src/gluonts/model/deepvar_hierarchical/_estimator.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import List, Optional
+
+import numpy as np
+from mxnet.gluon import HybridBlock
+import mxnet as mx
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.model.deepvar import DeepVAREstimator
+from gluonts.mx.model.predictor import Predictor
+from gluonts.mx.model.predictor import RepresentableBlockPredictor
+from gluonts.mx.distribution import DistributionOutput
+from gluonts.mx.trainer import Trainer
+from gluonts.mx.util import copy_parameters
+from gluonts.time_feature import TimeFeature
+from gluonts.transform import Transformation
+
+
+# Relative imports
+from ._network import (
+    DeepVARHierarchicalPredictionNetwork,
+    DeepVARHierarchicalTrainingNetwork,
+)
+
+
+def projection_mat(S, return_constraint_mat: bool = False):
+    """
+    Generates the projection matrix given the aggregation matrix `S`.
+
+    Parameters
+    ----------
+    S
+        Summation or aggregation matrix. Shape: (total_num_time_series, num_base_time_series)
+    return_constraint_mat
+        Return the coefficient matrix of the linear constraints?
+
+    Returns
+    -------
+    Tensor
+        Projection matrix, shape (total_num_time_series, total_num_time_series)
+    Tensor (if `return_constraint_mat` is True)
+        Coefficient matrix of the linear constraints, shape (num_agg_time_series, num_time_series)
+
+    """
+    # Re-arrange S matrix to form A matrix (coefficient matrix of the linear constraints)
+    # S = [S_agg|I_m_K]^T dim:(m,m_K)
+    # A = [I_magg | -S_agg] dim:(m_agg,m)
+
+    m, m_K = S.shape
+    m_agg = m - m_K
+
+    # The top `m_agg` rows of the matrix `S` give the aggregation constraint matrix.
+    S_agg = S[:m_agg, :]
+    A = np.hstack((np.eye(m_agg), -S_agg))
+
+    M = np.eye(m) - A.T @ np.linalg.pinv(A @ A.T) @ A
+
+    if return_constraint_mat:
```

Comment:
I'm not sure we should make the number of return objects dependent on a flag. This would cause that the downstream code does/does not work because it needs to take the flag into account. Could we just always return the `constraint_mat` and ignore it downstream if not used? Seems there is no option to set this in the estimator anyway.

Reply:
yes, anyway I need both, so returning both of them.